File: firstFollow.ml
(** FIRST and FOLLOW sets computation *)

open Utils
open Grammar

(** Type for FIRST sets *)
type first_sets = SymbolSet.t SymbolMap.t

(** Type for FOLLOW sets *)
type follow_sets = SymbolSet.t SymbolMap.t

(** Get FIRST set of a symbol, returning empty set if not found *)
let first_of_symbol first_map symbol =
  try SymbolMap.find symbol first_map
  with Not_found -> SymbolSet.empty

(** Get FOLLOW set of a symbol, returning empty set if not found *)
let follow_of_symbol follow_map symbol =
  try SymbolMap.find symbol follow_map
  with Not_found -> SymbolSet.empty

(** Compute FIRST set of a string (sequence of symbols)
    FIRST(X₁X₂...Xₙ) algorithm:
    - Add FIRST(X₁) - {ε} to result
    - If ε ∈ FIRST(X₁), add FIRST(X₂) - {ε}
    - Continue while ε ∈ FIRST(Xᵢ)
    - If ε ∈ FIRST(Xᵢ) for all i, add ε to result *)
let first_of_string first_map symbols =
  let rec aux result has_epsilon = function
    | [] ->
        if has_epsilon then SymbolSet.add Epsilon result
        else result
    | sym :: rest ->
        let first_sym = first_of_symbol first_map sym in
        let result' = SymbolSet.union result (SymbolSet.remove Epsilon first_sym) in
        if SymbolSet.mem Epsilon first_sym then
          aux result' true rest
        else
          result'
  in
  aux SymbolSet.empty false symbols

(** Compute FIRST sets for all symbols
    Algorithm (fixed-point iteration):
    1. For terminals: FIRST(a) = {a}
    2. For nonterminals A with production A → X₁X₂...Xₙ:
       - Add FIRST(X₁) - {ε} to FIRST(A)
       - If ε ∈ FIRST(X₁), add FIRST(X₂) - {ε}
       - Continue while ε ∈ FIRST(Xᵢ)
       - If ε ∈ FIRST(Xᵢ) for all i, add ε to FIRST(A)
    3. Repeat until no changes *)
let compute_first_sets grammar =
  let nonterminals = get_nonterminals grammar in
  let terminals = get_terminals grammar in
  let all_productions = get_all_productions grammar in

  (* Initialize FIRST sets *)
  let initial_first =
    (* Add terminals *)
    let with_terminals = SymbolSet.fold (fun t acc ->
      SymbolMap.add t (SymbolSet.singleton t) acc
    ) terminals SymbolMap.empty in
    (* Add epsilon *)
    let with_epsilon = SymbolMap.add Epsilon (SymbolSet.singleton Epsilon) with_terminals in
    (* Add end marker *)
    let with_end = SymbolMap.add EndMarker (SymbolSet.singleton EndMarker) with_epsilon in
    (* Initialize nonterminals with empty sets *)
    SymbolSet.fold (fun nt acc ->
      SymbolMap.add nt SymbolSet.empty acc
    ) nonterminals with_end
  in

  (* Fixed-point iteration *)
  let rec iterate first_map =
    let new_first_map = List.fold_left (fun acc_map prod ->
      let lhs = prod.lhs in
      let rhs = prod.rhs in
      let current_first = first_of_symbol acc_map lhs in

      (* Compute FIRST of RHS *)
      let rhs_first = first_of_string acc_map rhs in

      (* Union with current FIRST set *)
      let new_first = SymbolSet.union current_first rhs_first in

      SymbolMap.add lhs new_first acc_map
    ) first_map all_productions in

    (* Check if anything changed *)
    let changed = SymbolMap.exists (fun sym new_set ->
      let old_set = first_of_symbol first_map sym in
      not (SymbolSet.equal old_set new_set)
    ) new_first_map in

    if changed then iterate new_first_map
    else new_first_map
  in

  iterate initial_first

(** Compute FOLLOW sets for all nonterminals
    Algorithm (fixed-point iteration):
    1. FOLLOW(S) contains $
    2. For production A → αBβ:
       - Add FIRST(β) - {ε} to FOLLOW(B)
       - If ε ∈ FIRST(β) or β = ε, add FOLLOW(A) to FOLLOW(B)
    3. Repeat until no changes *)
let compute_follow_sets grammar first_map =
  let nonterminals = get_nonterminals grammar in
  let start_symbol = get_start_symbol grammar in
  let all_productions = get_all_productions grammar in

  (* Initialize FOLLOW sets *)
  let initial_follow =
    (* Initialize all nonterminals with empty sets *)
    let empty_follow = SymbolSet.fold (fun nt acc ->
      SymbolMap.add nt SymbolSet.empty acc
    ) nonterminals SymbolMap.empty in
    (* Add $ to FOLLOW(S) *)
    let start_follow = SymbolSet.singleton EndMarker in
    SymbolMap.add start_symbol start_follow empty_follow
  in

  (* Fixed-point iteration *)
  let rec iterate follow_map =
    let new_follow_map = List.fold_left (fun acc_map prod ->
      let lhs = prod.lhs in
      let rhs = prod.rhs in

      (* Process each position in the RHS *)
      let rec process_rhs acc beta = match beta with
        | [] -> acc
        | sym :: rest ->
            (* Only process nonterminals *)
            if not (is_nonterminal sym) then
              process_rhs acc rest
            else
              let current_follow = follow_of_symbol acc sym in

              (* Compute FIRST(rest) *)
              let first_rest = first_of_string first_map rest in

              (* Add FIRST(rest) - {ε} to FOLLOW(sym) *)
              let new_follow = SymbolSet.union current_follow
                              (SymbolSet.remove Epsilon first_rest) in

              (* If ε ∈ FIRST(rest) or rest is empty, add FOLLOW(lhs) to FOLLOW(sym) *)
              let new_follow' =
                if rest = [] || SymbolSet.mem Epsilon first_rest then
                  SymbolSet.union new_follow (follow_of_symbol acc lhs)
                else
                  new_follow
              in

              let acc' = SymbolMap.add sym new_follow' acc in
              process_rhs acc' rest
      in

      process_rhs acc_map rhs
    ) follow_map all_productions in

    (* Check if anything changed *)
    let changed = SymbolMap.exists (fun sym new_set ->
      let old_set = follow_of_symbol follow_map sym in
      not (SymbolSet.equal old_set new_set)
    ) new_follow_map in

    if changed then iterate new_follow_map
    else new_follow_map
  in

  iterate initial_follow

(** Print FIRST sets *)
let print_first_sets first_map grammar =
  print_endline "\nFIRST sets:";
  let nonterminals = get_nonterminals grammar in
  SymbolSet.iter (fun nt ->
    print_string ("FIRST(" ^ symbol_to_string nt ^ ") = ");
    print_symbol_set (first_of_symbol first_map nt)
  ) nonterminals

(** Print FOLLOW sets *)
let print_follow_sets follow_map grammar =
  print_endline "\nFOLLOW sets:";
  let nonterminals = get_nonterminals grammar in
  SymbolSet.iter (fun nt ->
    print_string ("FOLLOW(" ^ symbol_to_string nt ^ ") = ");
    print_symbol_set (follow_of_symbol follow_map nt)
  ) nonterminals

File: rust_version/cfg-parser/src/first_follow.rs
//! FIRST and FOLLOW set computation for context-free grammars.
//!
//! This module implements algorithms from Aho et al., "Compilers: Principles,
//! Techniques, and Tools" (2nd Edition), sections 4.4.

use crate::grammar::Grammar;
use crate::symbol::Symbol;
use std::collections::{HashMap, HashSet};

/// Type alias for FIRST sets mapping.
pub type FirstSets = HashMap<Symbol, HashSet<Symbol>>;

/// Type alias for FOLLOW sets mapping.
pub type FollowSets = HashMap<Symbol, HashSet<Symbol>>;

/// Computes the FIRST sets for all symbols in the grammar.
///
/// # Algorithm
/// 1. For terminals: FIRST(a) = {a}
/// 2. For nonterminals A with production A → X₁X₂...Xₙ:
///    - Add FIRST(X₁) - {ε} to FIRST(A)
///    - If ε ∈ FIRST(X₁), add FIRST(X₂) - {ε}
///    - Continue while ε ∈ FIRST(Xᵢ)
///    - If ε ∈ FIRST(Xᵢ) for all i, add ε to FIRST(A)
/// 3. Repeat until no changes (fixed-point iteration)
pub fn compute_first_sets(grammar: &Grammar) -> FirstSets {
    let mut first_sets: FirstSets = HashMap::new();

    // Initialize FIRST sets for terminals
    for terminal in grammar.terminals() {
        first_sets.insert(*terminal, HashSet::from([*terminal]));
    }

    // Initialize epsilon and end marker
    first_sets.insert(Symbol::Epsilon, HashSet::from([Symbol::Epsilon]));
    first_sets.insert(Symbol::EndMarker, HashSet::from([Symbol::EndMarker]));

    // Initialize nonterminals with empty sets
    for nonterminal in grammar.nonterminals() {
        first_sets.insert(*nonterminal, HashSet::new());
    }

    // Fixed-point iteration
    let mut changed = true;
    while changed {
        changed = false;

        for production in grammar.all_productions() {
            let lhs = production.lhs;
            let current_first = first_sets.get(&lhs).unwrap().clone();

            // Compute FIRST of RHS
            let rhs_first = first_of_string(&first_sets, &production.rhs);

            // Union with current FIRST set
            let new_first: HashSet<Symbol> = current_first.union(&rhs_first).copied().collect();

            if new_first.len() != current_first.len() {
                first_sets.insert(lhs, new_first);
                changed = true;
            }
        }
    }

    first_sets
}

/// Computes FIRST set of a string (sequence of symbols).
///
/// # Algorithm
/// - Add FIRST(X₁) - {ε} to result
/// - If ε ∈ FIRST(X₁), add FIRST(X₂) - {ε}
/// - Continue while ε ∈ FIRST(Xᵢ)
/// - If ε ∈ FIRST(Xᵢ) for all i, add ε to result
pub fn first_of_string(first_sets: &FirstSets, symbols: &[Symbol]) -> HashSet<Symbol> {
    let mut result = HashSet::new();
    let mut has_epsilon = true;

    for symbol in symbols {
        if !has_epsilon {
            break;
        }

        let first_sym = first_sets.get(symbol).cloned().unwrap_or_default();

        // Add FIRST(symbol) - {ε}
        for sym in &first_sym {
            if !sym.is_epsilon() {
                result.insert(*sym);
            }
        }

        // Check if epsilon is in FIRST(symbol)
        has_epsilon = first_sym.contains(&Symbol::Epsilon);
    }

    // If all symbols can derive epsilon, add epsilon to result
    if has_epsilon {
        result.insert(Symbol::Epsilon);
    }

    result
}

/// Computes the FOLLOW sets for all nonterminals in the grammar.
///
/// # Algorithm
/// 1. FOLLOW(S) contains $
/// 2. For production A → αBβ:
///    - Add FIRST(β) - {ε} to FOLLOW(B)
///    - If ε ∈ FIRST(β) or β = ε, add FOLLOW(A) to FOLLOW(B)
/// 3. Repeat until no changes (fixed-point iteration)
pub fn compute_follow_sets(grammar: &Grammar, first_sets: &FirstSets) -> FollowSets {
    let mut follow_sets: FollowSets = HashMap::new();

    // Initialize all nonterminals with empty sets
    for nonterminal in grammar.nonterminals() {
        follow_sets.insert(*nonterminal, HashSet::new());
    }

    // Add $ to FOLLOW(S)
    let start_symbol = grammar.start_symbol();
    follow_sets
        .get_mut(&start_symbol)
        .unwrap()
        .insert(Symbol::EndMarker);

    // Fixed-point iteration
    let mut changed = true;
    while changed {
        changed = false;

        for production in grammar.all_productions() {
            let lhs = production.lhs;
            let rhs = &production.rhs;

            // Process each position in the RHS
            for (i, symbol) in rhs.iter().enumerate() {
                // Only process nonterminals
                if !symbol.is_nonterminal() {
                    continue;
                }

                let current_follow = follow_sets.get(symbol).unwrap().clone();
                let mut new_follow = current_follow.clone();

                // Get the rest of the production after this symbol
                let beta = &rhs[i + 1..];

                // Compute FIRST(β)
                let first_beta = first_of_string(first_sets, beta);

                // Add FIRST(β) - {ε} to FOLLOW(symbol)
                for sym in &first_beta {
                    if !sym.is_epsilon() {
                        new_follow.insert(*sym);
                    }
                }

                // If ε ∈ FIRST(β) or β is empty, add FOLLOW(lhs) to FOLLOW(symbol)
                if beta.is_empty() || first_beta.contains(&Symbol::Epsilon) {
                    let follow_lhs = follow_sets.get(&lhs).unwrap().clone();
                    new_follow = new_follow.union(&follow_lhs).copied().collect();
                }

                if new_follow.len() != current_follow.len() {
                    follow_sets.insert(*symbol, new_follow);
                    changed = true;
                }
            }
        }
    }

    follow_sets
}
